File: test/go/widget/apply_ott_test.go
package widget_test

import (
	"context"
	"encoding/json"
	"fmt"
	"testing"
	"uat-script/helper"
	widget_helper "uat-script/widget"

	"github.com/dana-id/dana-go/v2/widget/v1"
)

const widgetApplyOttCase = "ApplyOtt"

// ApplyOtt
func TestApplyOttSuccess(t *testing.T) {
	helper.RetryTest(t, 3, 1, func() error {
		caseName := "ApplyOttSuccess"

		accessToken := obtainAccessTokenViaOAuth(t, helper.TestConfig.PhoneNumber, helper.TestConfig.PIN)
		fmt.Printf("Obtained access token: %s\n", accessToken)

		jsonDict, err := helper.GetRequest(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName)
		if err != nil {
			return fmt.Errorf("failed to get request data: %v", err)
		}

		if additionalInfo, ok := jsonDict["additionalInfo"].(map[string]interface{}); ok {
			additionalInfo["accessToken"] = accessToken
		} else {
			jsonDict["additionalInfo"] = map[string]interface{}{"accessToken": accessToken}
		}

		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			return fmt.Errorf("failed to marshal JSON: %v", err)
		}

		applyOttReq := widget.ApplyOTTRequest{}
		err = json.Unmarshal(jsonBytes, &applyOttReq)
		if err != nil {
			return fmt.Errorf("failed to unmarshal JSON: %v", err)
		}

		ctx := context.Background()
		apiResponse, httpResponse, err := helper.ApiClient.WidgetAPI.ApplyOTT(ctx).ApplyOTTRequest(applyOttReq).Execute()
		if err != nil {
			return fmt.Errorf("API request failed: %v", err)
		}
		defer httpResponse.Body.Close()

		responseJSON, err := apiResponse.MarshalJSON()
		if err != nil {
			return fmt.Errorf("failed to convert response to JSON: %v", err)
		}

		err = helper.AssertResponse(
			helper.TestConfig.JsonWidgetPath,
			widgetApplyOttCase,
			caseName,
			string(responseJSON),
			nil,
		)
		if err != nil {
			return err
		}
		return nil
	})
}

func TestApplyOttFailTokenNotFound(t *testing.T) {
	caseName := "ApplyOttCustomerTokenNotFound"
	jsonDict, err := helper.GetRequest(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName)
	if err != nil {
		t.Fatalf("Failed to get request data: %v", err)
	}

	// Only patch accessToken (matching PHP: $jsonDict['additionalInfo']['accessToken'] = 'invalid_access_token_for_testing')
	if additionalInfo, ok := jsonDict["additionalInfo"].(map[string]interface{}); ok {
		additionalInfo["accessToken"] = "invalid_access_token_for_testing"
	} else {
		jsonDict["additionalInfo"] = map[string]interface{}{"accessToken": "invalid_access_token_for_testing"}
	}

	jsonBytes, err := json.Marshal(jsonDict)
	if err != nil {
		t.Fatalf("Failed to marshal JSON: %v", err)
	}

	applyOttReq := widget.ApplyOTTRequest{}
	err = json.Unmarshal(jsonBytes, &applyOttReq)
	if err != nil {
		t.Fatalf("Failed to unmarshal JSON: %v", err)
	}

	ctx := context.Background()
	_, httpResponse, err := helper.ApiClient.WidgetAPI.ApplyOTT(ctx).ApplyOTTRequest(applyOttReq).Execute()
	if err != nil {
		err = helper.AssertFailResponse(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName, httpResponse, nil)
		if err != nil {
			t.Fatal(err)
		}
	} else {
		defer httpResponse.Body.Close()
		t.Fatalf("Expected ApiException was not thrown")
	}
}

func obtainAccessTokenViaOAuth(t *testing.T, phoneNumber, pin string) string {
	t.Helper()

	redirectURL, err := widget_helper.GetRedirectOauthUrl(phoneNumber, pin)
	if err != nil {
		t.Fatalf("Failed to get redirect OAuth URL: %v", err)
	}

	authCode, err := widget_helper.GetAuthCode(phoneNumber, pin, redirectURL)
	if err != nil {
		t.Fatalf("Failed to get auth code: %v", err)
	}

	ctx := context.Background()
	applyTokenReq := widget.NewApplyTokenAuthorizationCodeRequest("AUTHORIZATION_CODE", authCode)
	applyTokenReqValue := widget.ApplyTokenAuthorizationCodeRequestAsApplyTokenRequest(applyTokenReq)
	applyTokenResp, _, err := helper.ApiClient.WidgetAPI.ApplyToken(ctx).ApplyTokenRequest(applyTokenReqValue).Execute()
	if err != nil {
		t.Fatalf("Failed to obtain access token: %v", err)
	}

	return applyTokenResp.GetAccessToken()
}

func TestApplyOttCustomerAccountUserStatusAbnormal(t *testing.T) {
	helper.RetryTest(t, 3, 1, func() error {
		caseName := "ApplyOttCustomerAccountUserStatusAbnormal"

		accessToken := obtainAccessTokenViaOAuth(t, helper.TestConfig.AbnormalUserPhoneNumber, helper.TestConfig.AbnormalUserPIN)

		jsonDict, err := helper.GetRequest(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName)
		if err != nil {
			return fmt.Errorf("failed to get request data: %v", err)
		}

		if additionalInfo, ok := jsonDict["additionalInfo"].(map[string]interface{}); ok {
			additionalInfo["accessToken"] = accessToken
			additionalInfo["deviceId"] = helper.TestConfig.DeviceID
		} else {
			jsonDict["additionalInfo"] = map[string]interface{}{
				"accessToken": accessToken,
				"deviceId":    helper.TestConfig.DeviceID,
			}
		}

		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			return fmt.Errorf("failed to marshal JSON: %v", err)
		}

		applyOttReq := widget.ApplyOTTRequest{}
		if err = json.Unmarshal(jsonBytes, &applyOttReq); err != nil {
			return fmt.Errorf("failed to unmarshal JSON: %v", err)
		}

		ctx := context.Background()
		_, httpResponse, err := helper.ApiClient.WidgetAPI.ApplyOTT(ctx).ApplyOTTRequest(applyOttReq).Execute()
		if err != nil {
			if assertErr := helper.AssertFailResponse(helper.TestConfig.JsonWidgetPath, widgetApplyOttCase, caseName, httpResponse, nil); assertErr != nil {
				return assertErr
			}
			return nil
		}
		defer httpResponse.Body.Close()
		return fmt.Errorf("expected 4034905 error for case %s but API call succeeded", caseName)
	})
}

File: test/php/widget/ApplyOttTest.php
<?php

namespace DanaUat\Widget;

use PHPUnit\Framework\TestCase;
use Dana\Widget\v1\Api\WidgetApi;
use Dana\Configuration;
use Dana\ObjectSerializer;
use Dana\Env;
use Dana\ApiException;
use DanaUat\Helper\Assertion;
use DanaUat\Helper\Util;
use DanaUat\Widget\OauthUtil;
use Exception;
use GuzzleHttp\Psr7\Response;

class ApplyOttTest extends TestCase
{
    private static $titleCase = 'ApplyOtt';
    private static $jsonPathFile = 'resource/request/components/Widget.json';
    private static $apiInstance;
    private static $phoneNumber = '083811223355';
    private static $userPin = '181818';
    private static $deviceId = 'deviceid123';
    private static $authCode;
    private static $accessToken;

    public static function setUpBeforeClass(): void
    {
        $configuration = new Configuration();
        $configuration->setApiKey('PRIVATE_KEY', getenv('PRIVATE_KEY'));
        $configuration->setApiKey('ORIGIN', getenv('ORIGIN'));
        $configuration->setApiKey('X_PARTNER_ID', getenv('X_PARTNER_ID'));
        $configuration->setApiKey('ENV', Env::SANDBOX);
        self::$apiInstance = new WidgetApi(null, $configuration);
        
        // Get OAuth authorization code
        self::$authCode = OauthUtil::getAuthCode(
            getenv('X_PARTNER_ID'),
            null,
            self::$phoneNumber,
            self::$userPin,
            null
        );
        
        echo "\nObtained auth code: " . self::$authCode . "\n";
        
        // Get access token using the auth code
        $tokenCaseName = 'ApplyTokenSuccess';
        $tokenJsonDict = Util::getRequest(
            'resource/request/components/Widget.json',
            'ApplyToken',
            $tokenCaseName
        );
        
        $tokenRequestObj = ObjectSerializer::deserialize(
            $tokenJsonDict,
            'Dana\Widget\v1\Model\ApplyTokenRequest'
        );
        
        $tokenRequestObj->setAuthCode(self::$authCode);
        
        try {
            $apiResponse = self::$apiInstance->applyToken($tokenRequestObj);
            $responseJson = json_decode($apiResponse->__toString(), true);
            self::$accessToken = $responseJson['accessToken'];
            echo "\nObtained access token: " . self::$accessToken . "\n";
        } catch (Exception $e) {
            echo "\nFailed to obtain access token: " . $e->getMessage() . "\n";
        }
    }

    /**
     * Should give success response for apply ott scenario
     */
    public function testApplyOttSuccess(): void
    {

        Util::withDelay(function () {
            $caseName = 'ApplyOttSuccess';
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );
            $jsonDict['additionalInfo']['accessToken'] = self::$accessToken;
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );

            $apiResponse = self::$apiInstance->applyOTT($requestObj);
            
            Assertion::assertResponse(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName,
                $apiResponse->__toString()
            );
        });
    }

    /**
     * Should fail to apply OTT with token not found
     */
    public function testApplyOttFailTokenNotFound(): void
    {

        Util::withDelay(function () {
            $caseName = 'ApplyOttCustomerTokenNotFound';
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );
            // Set an invalid access token for testing token not found error
            $jsonDict['additionalInfo']['accessToken'] = 'invalid_access_token_for_testing';
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );
            try {
                self::$apiInstance->applyOTT($requestObj);
                $this->fail('Expected ApiException was not thrown');
            } catch (ApiException $e) {
                Assertion::assertFailResponse(self::$jsonPathFile, self::$titleCase, $caseName, $e->getResponseBody());
                $this->assertTrue(true);
            }
        });
    }

    /**
     * Should fail to apply OTT when customer account user status is abnormal (same as Go)
     */
    public function testApplyOttCustomerAccountUserStatusAbnormal(): void
    {
        Util::withDelay(function () {
            $caseName = 'ApplyOttCustomerAccountUserStatusAbnormal';
            $abnormalPhone = '0855100800';
            $abnormalPin = '146838';

            $authCode = OauthUtil::getAuthCode(
                getenv('X_PARTNER_ID'),
                null,
                $abnormalPhone,
                $abnormalPin,
                null
            );
            $this->assertNotEmpty($authCode, 'Failed to obtain auth code for abnormal user');

            $tokenJsonDict = Util::getRequest(
                'resource/request/components/Widget.json',
                'ApplyToken',
                'ApplyTokenSuccess'
            );
            $tokenRequestObj = ObjectSerializer::deserialize(
                $tokenJsonDict,
                'Dana\Widget\v1\Model\ApplyTokenRequest'
            );
            $tokenRequestObj->setAuthCode($authCode);
            $tokenResponse = self::$apiInstance->applyToken($tokenRequestObj);
            $accessToken = json_decode($tokenResponse->__toString(), true)['accessToken'];

            $jsonDict = Util::getRequest(self::$jsonPathFile, self::$titleCase, $caseName);
            $jsonDict['additionalInfo']['accessToken'] = $accessToken;
            $jsonDict['additionalInfo']['deviceId'] = self::$deviceId;
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );

            try {
                self::$apiInstance->applyOTT($requestObj);
                $this->fail('Expected ApiException was not thrown');
            } catch (ApiException $e) {
                Assertion::assertFailResponse(self::$jsonPathFile, self::$titleCase, $caseName, $e->getResponseBody());
                $this->assertTrue(true);
            }
        });
    }

    /**
     * @deprecated Wrong case name; use testApplyOttCustomerAccountUserStatusAbnormal
     */
    public function testApplyOttFailInvalidUserStatus(): void
    {
        $this->markTestSkipped('Replaced by testApplyOttCustomerAccountUserStatusAbnormal (same as Go)');

        Util::withDelay(function () {
            $caseName = 'ApplyOttFailInvalidUserStatus';
            $jsonDict = Util::getRequest(
                self::$jsonPathFile,
                self::$titleCase,
                $caseName
            );
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );
            // Set a modified access token that will trigger invalid user status error
            $jsonDict['additionalInfo']['accessToken'] = self::$accessToken !== null ? self::$accessToken.'_modified' : 'test_access_token';
            $requestObj = ObjectSerializer::deserialize(
                $jsonDict,
                'Dana\Widget\v1\Model\ApplyOTTRequest'
            );
            try {
                self::$apiInstance->applyOTT($requestObj);
                $this->fail('Expected ApiException was not thrown');
            } catch (ApiException $e) {
                Assertion::assertFailResponse(self::$jsonPathFile, self::$titleCase, $caseName, $e->getResponseBody());
                $this->assertTrue(true);
            }
        });
    }
}
